File: Algorithms/49.字母异位词分组/groupAnagrams.py
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        word_map = defaultdict(list)
        for s in strs:
            word_map[''.join(sorted(s))].append(s)
        return list(word_map.values())

File: 程序员面试金典/10.2/groupAnagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> wordMap = new HashMap<>();
        for (String s : strs) {
            String sorted = sort(s);
            if (!wordMap.containsKey(sorted)) {
                List<String> list = new ArrayList<>();
                list.add(s);
                wordMap.put(sorted, list);
            }
            else
                wordMap.get(sorted).add(s);
        }
        return new ArrayList<>(wordMap.values());
    }

    private String sort(String s) {
        char[] chars = s.toCharArray();
        Arrays.sort(chars);
        return new String(chars);
    }

}
